File: Arrays/03-remove_duplicates_sorted_array/remove_duplicates_sorted_array.py
from typing import List

# Brute Force Approach
# ---------------------
# INTUITION - use a set to store the seen numbers and if the number is not seen then add that into vector
# TIME COMPLEXITY - O(N)
# SPACE COMPLEXITY - O(2N) - set and result vector

def remove_duplicates_brute( array : List[int] ) -> List[int] :
    if not array or len(array) == 1: return array
    seen = set()
    res = []
    for number in array:
        if number not in seen:
            seen.add(number)
            res.append(number)

    return res

# Better Approach
# ---------------------
# INTUITION - create a new res array for answer, sort the given array and iterate over that, if the number not equal to the last number in the res array then push it in res array else pass.
# TIME COMPLEXITY - O(LogN + N) - O(LogN) to sort and O(N) to iterate over array to check for duplicates
# SPACE COMPLEXITY - O(N) - result vector to return, otherwise O(1)

def remove_duplicates_better( array : List[int] ) -> List[int] :
    if not array or len(array) == 1: return array
    array = sorted(array)
    res = []
    for number in array:
        if not res or res[-1] != number:
            res.append(number)

    return res

# Optimal Approach
# ---------------------
# INTUITION - Take one pointer back and iterate over the array, if the current number is not equal to the number at the back pointer index
# then increment the back index by 1 and make that number the current number else passon.
# TIME COMPLEXITY - O(N)
# SPACE COMPLEXITY - O(1)

def remove_duplicates_optimal( array : List[int] ) -> List[int] :
    if not array or len(array) == 1: return array
    back = 0
    for number in array[1:]:
        if array[back] != number:
            back+=1
            array[back] = number
    return array[:back+1]
        
        
# check
arr = [1, 1, 2, 2, 2, 3, 0, 4, 4, 4, 4, 0,4, 9, 9]
print(remove_duplicates_better(arr))

File: Arrays/03-remove_duplicates_sorted_array/remove_duplicates_sorted_array.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <unordered_set>
using namespace std;

/*
Brute Force Approach
---------------------
INTUITION - use a set to store the seen numbers and if the number is not seen then add that into vector
TIME COMPLEXITY - O(N)
SPACE COMPLEXITY - O(2N) - set and result vector

UNORDERED SET :-

Operation	            Average Case	   Worst Case
-------------------------------------------------------------------
Insertion (insert())	    O(1)	           O(N) (due to hash collisions)
Lookup (find(), count())	O(1)	           O(N)
Deletion (erase())	        O(1)               O(N)
Iteration (for x in set)	O(N)	           O(N)
*/

vector<int> remove_duplicates_brute(vector<int> array) {
    unordered_set<int> seen;
    vector<int> res;
    for (int num : array) {
        if (seen.find(num) == seen.end()) {
            seen.insert(num);
            res.push_back(num);
        }
    }
    return res;
}

/*
Better Approach
---------------------
INTUITION - create a new vector and iterate through the original vector, add into the result vector if the number is not equl to the last element of the vector
TIME COMPLEXITY - O(N)
SPACE COMPLEXITY - O(N)

*/

vector<int> remove_duplicates_better(vector<int> array) {
    vector<int> res;
    sort(res.begin(), res.end());
    for (int num : array) {
        if (res.empty() || num != res.back()) {
            res.push_back(num);
        }
    }
    return res;
}

/*
Optimal Approach
---------------------
Logic: Take a pointer back and initialize it to zero, loop from 1st index to end and if the number at that index is not equal to number at back index.
then make the number next to the back index that number and increment the back and at last resize the array to back + 1 cause the till back index all the 
number will be unique.

TIME COMPLEXITY - O(N)
SPACE COMPLEXITY - O(1)

*/

vector<int> remove_duplicates_optimal(vector<int> &array) {
    if (array.empty() || array.size() == 1) return array;
    
    int back = 0;
    
    for (int front = 1; front < array.size(); front++) {
        if (array[front] != array[back]) {
            back++;
            array[back] = array[front]; 
        }
    }
    array.resize(back + 1);
    return array;
}


int main() {    
    vector<int> vec = {1, 1, 2, 3, 5, 5, 7, 7, 7, 9, 9};
    // CHECK
    vector<int> res = remove_duplicates_optimal(vec);
    for (auto num: res){
        cout << num << " ";
    }cout << endl;
    return 0;
}